src/Database/Query.php

```diff
@@ -1328,6 +1330,32 @@ public function andHaving($conditions, $types = [])
         return $this;
     }
 
+    /**
+     * Adds a named window expression.
+     *
+     * Users are responsible for adding windows in the order each database requires.
+     *
+     * @param string $name Window name
+     * @param \Cake\Database\Expression\WindowExpression|\Closure $window Window expression
+     * @param bool $overwrite Clear all previous query window expressions
+     * @return $this
+     */
+    public function window(string $name, $window, bool $overwrite = false)
+    {
+        if ($overwrite) {
+            $this->_parts['window'] = [];
+        }
+
+        if ($window instanceof Closure) {
+            $window = $window(new WindowExpression(), $this);
```

Comment:
We should raise an error when the closure doesn't return a windowexpression.

Reply:
I didn't see other builders do this. Is that something we should change or does it happen elsewhere?

Reply:
Association `queryBuilder` callbacks raise a `RuntimeException` if the callback does not return a Query instance.